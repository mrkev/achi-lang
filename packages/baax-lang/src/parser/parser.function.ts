import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang } from "./sublang";
import { Node, withAt } from "./Node";

export type LangType_Function = {
  // printPoint(x: 3, y: 4)
  FunctionCall: Node<{
    kind: "FunctionCall";
    identifier: LangType["ValueIdentifier"];
    argument: LangType["RecordLiteral"];
  }>;

  // FunctionDefinition: {
  //   kind: "FunctionDefinition";
  //   argument: LangType["RecordDefinition"];
  //   body: LangType["Block"];
  // };

  // (x: number) => { ... }
  AnonymousFunctionLiteral: Node<{
    kind: "AnonymousFunctionLiteral";
    argument: LangType["RecordDefinition"];
    block: LangType["Block"];
  }>;
};

/**
 * Defines binary operations
 */
export const LangDef_Function = sublang<LangType, LangType_Function>({
  // log(msg: "hello")
  FunctionCall: (r) => {
    return withAt(
      Parsimmon.seqMap(
        r.ValueIdentifier,
        r.RecordLiteral,
        function (identifier, argument) {
          return {
            kind: "FunctionCall",
            identifier,
            argument,
          };
        }
      )
    );
  },

  // (foo: string) => {}
  AnonymousFunctionLiteral: (r) => {
    return withAt(
      Parsimmon.seqMap(
        r.RecordDefinition,
        r.__,
        Parsimmon.string("=>"),
        r.__,
        r.Block,
        (argument, _1, _2, _3, block) => {
          return {
            kind: "AnonymousFunctionLiteral",
            argument,
            block,
          };
        }
      )
    );
  },

  /* 
  TODO;
  const add = (x: number, y: number) => number {
  }

  function foo(x: number): number {
    return x + 3;
  }
   */
  // FunctionDefinition: (r) => {
  //   return Parsimmon.seqMap(
  //     r.RecordDefinition,
  //     r._,
  //     Parsimmon.string("=>"),
  //     // TODO: type tag?
  //     r._,
  //     r.Block,
  //     function (argument, _1, _2, _3, body) {
  //       return {
  //         kind: "FunctionDefinition",
  //         argument,
  //         body,
  //       };
  //     }
  //   );
  // },
});
