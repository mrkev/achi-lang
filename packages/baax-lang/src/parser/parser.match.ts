import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang, SublangTypedRule } from "./sublang";
import { Node, withAt } from "./Node";

export type LangType_Match = {
  CaseEntry: Node<{
    kind: "CaseEntry";
    guard: LangType["Expression"];
    block: LangType["Block"];
  }>;

  MatchFunction: Node<{
    kind: "MatchFunction";
    identifier: LangType["ValueIdentifier"];
    matchType: LangType["TypeExpression"];
    block: LangType["BlockOfCases"];
  }>;

  MatchExpression: Node<{
    kind: "MatchExpression";
    expression: LangType["Expression"];
    block: LangType["BlockOfCases"];
  }>;

  BlockOfCases: Node<{
    kind: "BlockOfCases";
    caseEntries: Array<LangType["CaseEntry"]>;
  }>;
};

/**
 * Defines binary operations
 */
export const LangDef_Match: SublangTypedRule<LangType, LangType_Match> =
  sublang<LangType, LangType_Match>({
    // case (card: King(value)): { ... }
    CaseEntry: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("case"),
          r.__,
          r.Expression,
          r._,
          Parsimmon.string(":"),
          r.__,
          r.Block,
          function (_0, _1, expression, _3, _4, _5, block) {
            return {
              kind: "CaseEntry",
              guard: expression,
              block,
            };
          }
        )
      );
    },

    // function foo matches Type {}
    MatchFunction: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("function"),
          r.__,
          r.ValueIdentifier,
          r.__,
          Parsimmon.string("matches"),
          r.__,
          r.TypeExpression,
          r._,
          r.BlockOfCases,
          function (_0, _1, identifier, _3, _4, _5, type, _7, block) {
            return {
              kind: "MatchFunction",
              identifier,
              matchType: type,
              block,
            };
          }
        )
      );
    },

    // match (card) { case King(): ... }
    MatchExpression: (r) => {
      return withAt(
        Parsimmon.seqMap(
          Parsimmon.string("match"),
          r._,
          Parsimmon.string("("),
          r._,
          r.Expression,
          r._,
          Parsimmon.string(")"),
          r._,
          r.BlockOfCases,
          function (_0, _1, _2, _3, expression, _5, _6, _7, block) {
            return {
              kind: "MatchExpression",
              expression,
              block,
            };
          }
        )
      );
    },

    BlockOfCases: (r) => {
      const caseListParser = Parsimmon.sepBy(r.CaseEntry, r.__nl).trim(r._nl);
      return withAt(
        Parsimmon.string("{")
          // TBD: sepBy1 and make a different parser for Unit?
          .then(caseListParser)
          .skip(Parsimmon.string("}"))
          .map((caseEntries) => {
            return {
              kind: "BlockOfCases",
              caseEntries,
            };
          })
      );
    },
  });
