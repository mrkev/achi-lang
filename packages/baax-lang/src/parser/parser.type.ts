import * as Parsimmon from "parsimmon";
import { Node } from "./Node";
import { LangType } from "./parser";
import { sublang } from "./sublang";

type PrefixUnaryTypeOperator =
  // for numbers
  | "-"
  // to make readonly
  | "+";
export type PrefixUnaryTypeOperation = Node<{
  kind: "PrefixUnaryTypeOperation";
  operator: PrefixUnaryTypeOperator;
  value: NEXT_PARSER;
}>;

type BinaryTypeOperator = "|" | "&";
export type BinaryTypeOperation = Node<{
  kind: "BinaryTypeOperation";
  operator: BinaryTypeOperator;
  left: NEXT_PARSER;
  right: NEXT_PARSER;
}>;

export type OperatableTypeExpression =
  | LangType["NumberLiteral"]
  | LangType["TypeIdentifier"]
  | LangType["StringLiteral"]
  | LangType["BooleanLiteral"]
  | LangType["RecordDefinition"];

type NEXT_PARSER =
  | OperatableTypeExpression
  | PrefixUnaryTypeOperation
  | BinaryTypeOperation;

// Takes a parser for the prefix operator, and a parser for the base thing being
// parsed, and parses as many occurrences as possible of the prefix operator.
// Note that the parser is created using `Parsimmon.lazy` because it's recursive. It's
// valid for there to be zero occurrences of the prefix operator.
function PREFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<PrefixUnaryTypeOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const res: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.seqMap(
      Parsimmon.index,
      operatorsParser,
      parser,
      Parsimmon.index,
      (start, operator, value, end) =>
        ({
          kind: "PrefixUnaryTypeOperation",
          operator: operator.op,
          value,
          "@": { start, end },
        } as const)
    ).or(nextParser);
    return res;
  });
  return parser;
}

type OpLoc<T extends string> = {
  op: T;
  start: Parsimmon.Index;
  end: Parsimmon.Index;
};

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// left. (e.g. 1-2-3 is (1-2)-3 not 1-(2-3))
function BINARY_LEFT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryTypeOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // We run into a similar problem as with the `POSTFIX` parser above where we
  // can't recurse in the direction we want, so we have to resort to parsing an
  // entire list of operator chunks and then using `.reduce` to manually nest
  // them again.
  //
  // Example:
  //
  // INPUT  :: "1+2+3"
  // PARSE  :: [1, ["+", 2], ["+", 3]]
  // REDUCE :: ["+", ["+", 1, 2], 3]
  return Parsimmon.seqMap(
    nextParser,
    Parsimmon.seq(operatorsParser, nextParser).many(),
    (first, rest) => {
      return rest.reduce((acc, ch) => {
        const [op, another] = ch;
        return {
          kind: "BinaryTypeOperation",
          operator: op.op,
          left: acc,
          right: another,
          "@": {
            start: acc["@"].start,
            end: another["@"].end,
          },
        } as const;
      }, first);
    }
  );
}

export function OperatorParser(r: Parsimmon.TypedLanguage<LangType>) {
  // Operators should allow whitespace around them, but not require it. This
  // helper combines multiple operators together with names.
  //
  // Example: operators({Add: "+", Sub: "-"})
  //
  // Gives back an operator that parses either + or - surrounded by optional
  // whitespace, and gives back the word "Add" or "Sub" instead of the character.
  function operators<Ops extends readonly string[]>(
    ops: Ops
  ): Parsimmon.Parser<OpLoc<Ops[number]>> {
    const withWhitespace = ops.map((sym) =>
      Parsimmon.seqMap(
        Parsimmon.index,
        r._,
        Parsimmon.string(sym),
        r._,
        Parsimmon.index,
        (start, _0, op, _2, end) => {
          return { op, start, end };
        }
      )
    );
    const res = Parsimmon.alt(...withWhitespace);
    return res;
  }

  // A basic value is any parenthesized expression or a number.
  // TODO: start, end don't take the parens into consideration. Is that an issue?
  const Basic: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.lazy<NEXT_PARSER>(() =>
    Parsimmon.string("(")
      .then(MyMath)
      .skip(Parsimmon.string(")"))
      .or(r.RecordDefinition)
      .or(r.NumberLiteral)
      .or(r.StringLiteral)
      .or(r.BooleanLiteral)
      .or(r.TypeIdentifier)
  );

  // Now we can describe the operators in order by precedence. You just need to
  // re-order the table.
  // const table = [
  //   { type: POSTFIX, ops: operators(["!"] as const) },
  //   { type: PREFIX, ops: operators(["-", "!"] as const) },
  //   { type: BINARY_RIGHT, ops: operators(["^"] as const) },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["*", "/"] as const),
  //   },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["+", "-"] as const),
  //   },
  // ];

  const table: Array<
    (next: Parsimmon.Parser<NEXT_PARSER>) => Parsimmon.Parser<NEXT_PARSER>
  > = [
    (next) => PREFIX(operators(["-", "+"] as const), next),
    (next) => BINARY_LEFT(operators(["&", "|"] as const), next),
  ];

  // Start off with Num as the base parser for numbers and thread that through the
  // entire table of operator parsers.
  const tableParser = table.reduce((next, level) => level(next), Basic);

  // The above is equivalent to:
  //
  // TYPE(operators({...}),
  //   TYPE(operators({...}),
  //     TYPE(operators({...})),
  //       TYPE(operators({...}),
  //         TYPE(operators({...}), ...))))
  //
  // But it's easier if to see what's going on and reorder the precedence if we
  // keep it in a table instead of nesting it all manually.

  // This is our version of a math expression.
  const MyMath = tableParser.trim(r._);
  return MyMath;
}

export type LangType_Type = {
  TypeExpression:
    | OperatableTypeExpression
    | PrefixUnaryTypeOperation
    | BinaryTypeOperation;

  // :string
  TypeTag: Node<{
    kind: "TypeTag";
    optional: boolean;
    typeExpression: LangType["TypeExpression"];
  }>;
};

/**
 * Defines type expressions and operations
 */
export const LangDef_Type = sublang<LangType, LangType_Type>({
  TypeExpression: OperatorParser,

  // :string
  TypeTag: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.index,
      Parsimmon.alt(Parsimmon.string("?:"), Parsimmon.string(":")),
      r._,
      r.TypeExpression,
      Parsimmon.index,
      (start, tagger, _3, typeExpression, end) => {
        const optional = tagger === "?:";
        return {
          kind: "TypeTag",
          optional,
          typeExpression,
          "@": { start, end },
        };
      }
    );
  },
});
