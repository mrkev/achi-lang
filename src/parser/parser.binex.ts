import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { Node, withAt } from "./Node";

type PrefixUnaryOperator = "-" | "!";
export type PrefixUnaryOperation = Node<{
  kind: "PrefixUnaryOperation";
  operator: PrefixUnaryOperator;
  value: NEXT_PARSER;
}>;

type SuffixUnaryOperator = "!";
export type SuffixUnaryOperation = Node<{
  kind: "SuffixUnaryOperation";
  operator: SuffixUnaryOperator;
  value: NEXT_PARSER;
}>;

type BinaryOperator =
  // String + Numerial
  | "+"
  // Numerical
  | "-"
  | "*"
  | "/"
  | "^"
  // Boolean
  | "&&"
  | "||"
  // Comparison
  | ">"
  | "<"
  | ">="
  | "<="
  | "=="
  | "!=";

export type BinaryOperation = Node<{
  kind: "BinaryOperation";
  operator: BinaryOperator;
  left: NEXT_PARSER;
  right: NEXT_PARSER;
}>;

export type OperatableExpression =
  | LangType["NumberLiteral"]
  | LangType["ValueIdentifier"]
  | LangType["StringLiteral"]
  | LangType["BooleanLiteral"]
  | LangType["FunctionCall"]
  | LangType["NullLiteral"];

// This parser supports basic math with + - * / ^, unary negation, factorial,
// and parentheses. It does not evaluate the math, just turn it into a series of
// nested lists that are easy to evaluate.

// You might think that parsing math would be easy since people learn it early
// in school, but dealing with precedence and associativity of operators is
// actually one of the hardest and most tedious things you can do in a parser!
// If you look at a language like JavaScript, it has even more operators than
// math, like = and && and || and ++ and so many more...

///////////////////////////////////////////////////////////////////////

type NEXT_PARSER =
  | OperatableExpression
  | PrefixUnaryOperation
  | SuffixUnaryOperation
  | BinaryOperation;

// Takes a parser for the prefix operator, and a parser for the base thing being
// parsed, and parses as many occurrences as possible of the prefix operator.
// Note that the parser is created using `Parsimmon.lazy` because it's recursive. It's
// valid for there to be zero occurrences of the prefix operator.
function PREFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<PrefixUnaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const res: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.seqMap(
      Parsimmon.index,
      operatorsParser,
      parser,
      Parsimmon.index,
      (start, operator, value, end) =>
        ({
          kind: "PrefixUnaryOperation",
          operator: operator.op,
          value,
          "@": { start, end },
        } as const)
    ).or(nextParser);
    return res;
  });
  return parser;
}

type OpLoc<T extends string> = {
  op: T;
  start: Parsimmon.Index;
  end: Parsimmon.Index;
};

// Ideally this function would be just like `PREFIX` but reordered like
// `Parsimmon.seq(parser, operatorsParser).or(nextParser)`, but that doesn't work. The
// reason for that is that Parsimmon will get stuck in infinite recursion, since
// the very first rule. Inside `parser` is to match parser again. Alternatively,
// you might think to try `nextParser.or(Parsimmon.seq(parser, operatorsParser))`, but
// that won't work either because in a call to `.or` (aka `Parsimmon.alt`), Parsimmon
// takes the first possible match, even if subsequent matches are longer, so the
// parser will never actually look far enough ahead to see the postfix
// operators.
function POSTFIX(
  operatorsParser: Parsimmon.Parser<OpLoc<"!">>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // Because we can't use recursion like stated above, we just match a flat list
  // of as many occurrences of the postfix operator as possible, then use
  // `.reduce` to manually nest the list.
  //
  // Example:
  //
  // INPUT  :: "4!!!"
  // PARSE  :: [4, "factorial", "factorial", "factorial"]
  // REDUCE :: ["factorial", ["factorial", ["factorial", 4]]]
  return Parsimmon.seqMap(nextParser, operatorsParser.many(), (x, suffixes) =>
    suffixes.reduce((acc, { op, end }) => {
      return {
        kind: "SuffixUnaryOperation",
        operator: op,
        value: acc,
        "@": {
          start: acc["@"].start,
          end,
        },
      } as const;
    }, x)
  );
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// right. (e.g. 1^2^3 is 1^(2^3) not (1^2)^3)
function BINARY_RIGHT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
): Parsimmon.Parser<NEXT_PARSER> {
  const parser = Parsimmon.lazy<NEXT_PARSER>(() => {
    const chainRes: Parsimmon.Parser<NEXT_PARSER> = nextParser.chain((next) =>
      withAt(
        Parsimmon.seqMap(
          operatorsParser,
          Parsimmon.of(next),
          parser,
          (operator, next, p) =>
            ({
              kind: "BinaryOperation",
              operator: operator.op,
              left: next,
              right: p,
            } as const)
        )
      ).or(Parsimmon.of(next))
    );

    return chainRes;
  });
  return parser;
}

// Takes a parser for all the operators at this precedence level, and a parser
// that parsers everything at the next precedence level, and returns a parser
// that parses as many binary operations as possible, associating them to the
// left. (e.g. 1-2-3 is (1-2)-3 not 1-(2-3))
function BINARY_LEFT(
  operatorsParser: Parsimmon.Parser<OpLoc<BinaryOperator>>,
  nextParser: Parsimmon.Parser<NEXT_PARSER>
) {
  // We run into a similar problem as with the `POSTFIX` parser above where we
  // can't recurse in the direction we want, so we have to resort to parsing an
  // entire list of operator chunks and then using `.reduce` to manually nest
  // them again.
  //
  // Example:
  //
  // INPUT  :: "1+2+3"
  // PARSE  :: [1, ["+", 2], ["+", 3]]
  // REDUCE :: ["+", ["+", 1, 2], 3]
  return Parsimmon.seqMap(
    nextParser,
    Parsimmon.seq(operatorsParser, nextParser).many(),
    (first, rest) => {
      return rest.reduce((acc, ch) => {
        const [op, another] = ch;
        return {
          kind: "BinaryOperation",
          operator: op.op,
          left: acc,
          right: another,
          "@": {
            start: acc["@"].start,
            end: another["@"].end,
          },
        } as const;
      }, first);
    }
  );
}

export function OperatorParser(r: Parsimmon.TypedLanguage<LangType>) {
  // Operators should allow whitespace around them, but not require it. This
  // helper combines multiple operators together with names.
  //
  // Example: operators({Add: "+", Sub: "-"})
  //
  // Gives back an operator that parses either + or - surrounded by optional
  // whitespace, and gives back the word "Add" or "Sub" instead of the character.
  function operators<Ops extends readonly string[]>(
    ops: Ops
  ): Parsimmon.Parser<OpLoc<Ops[number]>> {
    const withWhitespace = ops.map((sym) =>
      Parsimmon.seqMap(
        Parsimmon.index,
        r._,
        Parsimmon.string(sym),
        r._,
        Parsimmon.index,
        (start, _0, op, _2, end) => {
          return { op, start, end };
        }
      )
    );
    const res = Parsimmon.alt(...withWhitespace);
    return res;
  }

  // A basic value is any parenthesized expression or a number.
  // TODO: start, end don't take the parens into consideration. Is that an issue?
  const Basic: Parsimmon.Parser<NEXT_PARSER> = Parsimmon.lazy<NEXT_PARSER>(() =>
    Parsimmon.string("(")
      .then(MyMath)
      .skip(Parsimmon.string(")"))
      .or(r.NumberLiteral)
      .or(r.StringLiteral)
      .or(r.BooleanLiteral)
      .or(r.FunctionCall)
      .or(r.NullLiteral)
      .or(r.ValueIdentifier)
  );

  // Now we can describe the operators in order by precedence. You just need to
  // re-order the table.
  // const table = [
  //   { type: POSTFIX, ops: operators(["!"] as const) },
  //   { type: PREFIX, ops: operators(["-", "!"] as const) },
  //   { type: BINARY_RIGHT, ops: operators(["^"] as const) },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["*", "/"] as const),
  //   },
  //   {
  //     type: BINARY_LEFT,
  //     ops: operators(["+", "-"] as const),
  //   },
  // ];

  const table: Array<
    (next: Parsimmon.Parser<NEXT_PARSER>) => Parsimmon.Parser<NEXT_PARSER>
  > = [
    (next) => BINARY_LEFT(operators(["==", "!="] as const), next),
    (next) => POSTFIX(operators(["!"] as const), next),
    (next) => PREFIX(operators(["-", "!"] as const), next),
    (next) => BINARY_RIGHT(operators(["^"] as const), next),
    (next) => BINARY_LEFT(operators(["*", "/"] as const), next),
    (next) => BINARY_LEFT(operators(["+", "-"] as const), next),
    (next) => BINARY_LEFT(operators(["&&"] as const), next),
    (next) => BINARY_LEFT(operators(["||"] as const), next),
    (next) => BINARY_LEFT(operators([">=", "<=", ">", "<"] as const), next),
  ];

  // Start off with Num as the base parser for numbers and thread that through the
  // entire table of operator parsers.
  const tableParser = table.reduce((next, level) => level(next), Basic);

  // The above is equivalent to:
  //
  // TYPE(operators({...}),
  //   TYPE(operators({...}),
  //     TYPE(operators({...})),
  //       TYPE(operators({...}),
  //         TYPE(operators({...}), ...))))
  //
  // But it's easier if to see what's going on and reorder the precedence if we
  // keep it in a table instead of nesting it all manually.

  // This is our version of a math expression.
  const MyMath = tableParser.trim(r._);
  return MyMath;
}
