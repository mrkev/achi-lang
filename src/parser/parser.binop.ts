import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { MathParser, NEXT_PARSER } from "./parser.binex";
import { sublang } from "./sublang";

// expression =
//     term
//     | expression "+" term
//     | expression "-" term .
// term =
//     factor
//     | term "*" factor
//     | term "/" factor .
// factor =
//     number
//     | "(" expression ")" .

export type BinaryExpression = {
  kind: "BinaryExpression";
  operator: string;
  left: LangType["Expression"];
  right: LangType["Expression"];
};

export type LangType_BinOp = {
  // BinaryOperator: string;
  // BinaryOperation: {
  //   kind: "BinaryOperation";
  //   operator: string;
  //   left: LangType["Expression"];
  //   right: LangType["Expression"];
  // };

  // BinaryExpression_Expr
  // BinaryExpression: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Term: BinaryExpression | LangType["Expression"];
  // BinaryExpression_Factor: BinaryExpression | LangType["Expression"];
  OperatorExpression: NEXT_PARSER;
};

/**
 * Defines binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperatorExpression: MathParser,

  // BinaryExpression: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Term,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression,
  //       r._,
  //       Parsimmon.regex(/(\+|\-)/),
  //       r._,
  //       r.BinaryExpression_Term,
  //       (expr, _1, op, _3, term): LangType["BinaryExpression"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: expr,
  //           operator: op,
  //           right: term,
  //         };
  //       }
  //     )
  //   );
  // },

  // // // term =
  // //     factor
  // //     | term "*" factor
  // //     | term "/" factor .
  // BinaryExpression_Term: (r) => {
  //   return Parsimmon.alt(
  //     r.BinaryExpression_Factor,
  //     Parsimmon.seqMap(
  //       r.BinaryExpression_Term,
  //       r._,
  //       Parsimmon.regex(/(\*|\/)/),
  //       r._,
  //       r.BinaryExpression_Factor,
  //       (term, _1, op, _3, factor): LangType["BinaryExpression_Term"] => {
  //         return {
  //           kind: "BinaryExpression",
  //           left: term,
  //           operator: op,
  //           right: factor,
  //         };
  //       }
  //     )
  //   );
  // },

  // BinaryExpression_Factor: (r) => {
  //   return Parsimmon.seqMap(
  //     Parsimmon.string("<<"),
  //     r._,
  //     r.BinaryExpression,
  //     r._,
  //     Parsimmon.string(">>"),
  //     (_0, _1, expr, _3, _4) => {
  //       return expr;
  //     }
  //   ).or(r.NumberLiteral);
  // },

  // Arithmetic: + - / * %
  // Comparison: all double << >> == !=
  // Logic: || &&
  // Todo, bitwise?
  // BinaryOperator: () => {
  //   return Parsimmon.regex(/(\+|\-|\/|\*|\%|\=\=|\!\=|\>\>|\<\<|\|\||\&\&)/);
  // },

  // 3 + 3
  // BinaryOperation: (r) => {
  //   return Parsimmon.seqMap(
  //     r.Expression,
  //     r._,
  //     r.BinaryOperator,
  //     r._,
  //     r.Expression,
  //     (left, _0, operator, _1, right) => {
  //       return {
  //         kind: "BinaryOperation",
  //         operator,
  //         left,
  //         right,
  //       };
  //     }
  //   );
  // },
});

// Factor
// = "(" _ expr:Expression _ ")" { return expr; }
// / Integer
