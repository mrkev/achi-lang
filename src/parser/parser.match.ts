import * as Parsimmon from "parsimmon";
import { LangType } from "./parser";
import { sublang } from "./sublang";

export type LangType_Match = {
  CaseEntry: {
    kind: "CaseEntry";
    guard: LangType["Expression"];
    block: LangType["Block"];
  };

  MatchFunction: {
    kind: "MatchFunction";
    identifier: LangType["ValueIdentifier"];
    matchType: LangType["Type"];
    block: LangType["BlockOfCases"];
  };

  BlockOfCases: {
    kind: "BlockOfCases";
    caseEntries: Array<LangType["CaseEntry"]>;
  };
};

/**
 * Defines binary operations
 */
export const LangDef_Match = sublang<LangType, LangType_Match>({
  // case (card: King(value)): { ... }
  CaseEntry: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("case"),
      r.__,
      r.Expression,
      r._,
      Parsimmon.string(":"),
      r.__,
      r.Block,
      function (_0, _1, expression, _3, _4, _5, block) {
        return {
          kind: "CaseEntry",
          guard: expression,
          block,
        };
      }
    );
  },

  // function foo matches Type {}
  MatchFunction: (r) => {
    return Parsimmon.seqMap(
      Parsimmon.string("function"),
      r.__,
      r.ValueIdentifier,
      r.__,
      Parsimmon.string("matches"),
      r.__,
      r.Type,
      r.__,
      r.BlockOfCases,
      function (_0, _1, identifier, _3, _4, _5, type, _7, block) {
        return {
          kind: "MatchFunction",
          identifier,
          matchType: type,
          block: block,
        };
      }
    );
  },

  BlockOfCases: (r) => {
    const caseListParser = Parsimmon.sepBy(r.CaseEntry, r.__nl).trim(r._nl);

    return (
      Parsimmon.string("{")
        // TBD: sepBy1 and make a different parser for Unit?
        .then(caseListParser)
        .skip(Parsimmon.string("}"))
        .map((caseEntries) => {
          return {
            kind: "BlockOfCases",
            caseEntries,
          };
        })
    );
  },
});
