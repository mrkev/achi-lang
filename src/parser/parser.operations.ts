import { LangType } from "./parser";
import {
  BinaryOperation,
  OperatableExpression,
  OperatorParser,
  PrefixUnaryOperation,
  SuffixUnaryOperation,
} from "./parser.binex";
import { sublang } from "./sublang";

export type LangType_BinOp = {
  OperationExpression:
    | OperatableExpression
    | PrefixUnaryOperation
    | SuffixUnaryOperation
    | BinaryOperation;
};

/**
 * Defines unary and binary operations
 */
export const LangDef_BinOp = sublang<LangType, LangType_BinOp>({
  OperationExpression: OperatorParser,
});
