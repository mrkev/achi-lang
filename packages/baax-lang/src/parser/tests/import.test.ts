import { Lang } from "../parser";

test("ImportStatement.1", () => {
  const result = Lang.ImportStatement.tryParse('import Foo from "./bar"');
  expect(result).toMatchSnapshot();
});

test("ImportStatement.1", () => {
  const result = Lang.Program.tryParse('import Foo from "./bar"');
  expect(result).toMatchSnapshot();
});
