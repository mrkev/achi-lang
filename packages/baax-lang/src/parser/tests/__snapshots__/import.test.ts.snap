// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportStatement.1 1`] = `
{
  "@": {
    "end": {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "identifier": {
    "@": {
      "end": {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "kind": "AnyIdentifier",
    "value": "Foo",
  },
  "kind": "ImportStatement",
  "src": {
    "@": {
      "end": {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
      "start": {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
    },
    "kind": "StringLiteral",
    "value": "./bar",
  },
}
`;

exports[`ImportStatement.1 2`] = `
{
  "@": {
    "end": {
      "column": 24,
      "line": 1,
      "offset": 23,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "kind": "Program",
  "statements": [
    {
      "@": {
        "end": {
          "column": 24,
          "line": 1,
          "offset": 23,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "identifier": {
        "@": {
          "end": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": {
            "column": 8,
            "line": 1,
            "offset": 7,
          },
        },
        "kind": "AnyIdentifier",
        "value": "Foo",
      },
      "kind": "ImportStatement",
      "src": {
        "@": {
          "end": {
            "column": 24,
            "line": 1,
            "offset": 23,
          },
          "start": {
            "column": 17,
            "line": 1,
            "offset": 16,
          },
        },
        "kind": "StringLiteral",
        "value": "./bar",
      },
    },
  ],
}
`;
